Articles/Blog/SoftwareVariants.md

```diff
@@ -0,0 +1,165 @@
+
+## HPC Examples
+
+The most common pattern for HPC software are the options for parallelism.
+In the simplest case, this is a flag that enables or disables MPI support for the software.
+With the rise of GPU-computing over the last decade, a similar option is often added for GPU acceleration.
+
+Library and application developers in HPC might then add further more domain-specific compilation options, such as:
+- additional numerical solvers supported by external math libraries, e.g., BLAS/LAPACK/FFT
+- choices of geometry of a model
+
+## Motivation / The Catch (multiple)
+
+Binary options can be as colorful as the creativity of the developer that writes them.
+There are generally two options: boolean switches and multi-variant options.
+Binary switches are simply turning a functionality on/off, e.g., MPI support.
+Multi-variant options might be more involved, e.g., accelerating code for a specific GPU programming environment (CUDA/HIP/SYCL/OpenMP or OpenACC offloading).
+
+The catch with such options lies in sharing software with other people that like to re-use it in their own libraries and applications, develop against it and deploy their own product further to other developers and users.
+Problems for "downstream" depending developers of HPC software arise mainly from the implementation of these options.
+
+For instance, runtime options that change with the choosen binary option need to be documented to users/downstream developers - and complicate user experience in already tricky installations.
+Workflows have to be established when switching functionality downstream: change the software itself and install, find and use another variant of upstream packages.
+Testing gets more complicated: full recompilations are needed of vast parts of the software stack if one cannot create a single environment that enables and tests all functionality, because it might conflict.
+Even with sufficient automation, the latter increases continuous integration time and deployment resources.
+
+Typical problems shall be illustrated in a few examples.
+
+### Breaking Public API
+
+A common issues from quickly scattered `#ifdef` switches are changing API interfaces: extra parameters in functions, different class signatures and constructors, varying members.
+In the worst case, the change of the upstream option has to be mirrored 1:1 downstream with `#ifdef`s at call locations, e.g., because existing signatures are *changed*.
+
+Such API changes also always lead to incompatible application binary interfaces (ABIs).
+Likewise, it is easy to introduce breaking ABI changes that are more subtle, which will result in linkage failures for a new variant at best and subtle breakage at runtime in the worst case.
+A typical example is the manipulation of the number of member variables of a public class based on a binary variant.
+
```

Comment:
I can add a section here on actual ABI breaks, e.g., adding a private member to a public class at the wrong location

Reply:
done